src/macros.rs

```diff
@@ -0,0 +1,53 @@
+// Various macros for implmenting repetative pass-by-value semantics
+
+#[macro_export]
+macro_rules! add_impl {
```

Comment:
Maybe we should extend these macros to also generate implementations of "mixed" operators like `&FieldElement * FieldElement` and `FieldElement * &FieldElement`, so that all cases of borrow / non-borrow are covered?